Applications/FileIO/SWMM/SWMMInterface.cpp

```diff
@@ -0,0 +1,1248 @@
+}
+
+std::vector<double> SwmmInterface::getArrayForObject(SwmmObject obj_type, std::size_t obj_idx, std::size_t var_idx) const
+{
+    std::vector<double> data;
+    std::string const outfile (_base_name + ".out");
+    if (OpenSwmmOutFile(const_cast<char*>(outfile.c_str())) != 0)
+        return data;
+
+    bool is_var_idx_okay = true;
+    bool is_obj_idx_okay = true;
+    int obj_type_id;
+    switch (obj_type)
+    {
+        case SwmmObject::SUBCATCHMENT:
+            obj_type_id = 0;
+            if (obj_idx >= SWMM_Nsubcatch)
+                is_obj_idx_okay = false;
+            if (var_idx > (n_obj_params[obj_type_id] + SWMM_Npolluts))
+                is_var_idx_okay = false;
+            break;
+        case SwmmObject::NODE:
+            obj_type_id = 1;
+            if (obj_idx >= SWMM_Nnodes)
+                is_obj_idx_okay = false;
+            if (var_idx > (n_obj_params[obj_type_id] + SWMM_Npolluts))
+                is_var_idx_okay = false;
+            break;
+        case SwmmObject::LINK:
+            obj_type_id = 2;
+            if (obj_idx >= SWMM_Nlinks)
+                is_obj_idx_okay = false;
+            if (var_idx > (n_obj_params[obj_type_id] + SWMM_Npolluts))
+                is_var_idx_okay = false;
+            break;
+        case SwmmObject::SYSTEM:
+            obj_type_id = 3;
+            if (obj_idx >= 1)
+                is_obj_idx_okay = false;
+            if (var_idx > n_obj_params[obj_type_id])
+                is_var_idx_okay = false;
+            break;
+      default:
+         ERR ("Object type not recognised.");
+         CloseSwmmOutFile();
+         return data;
+    }
+
+    if (!is_obj_idx_okay)
+    {
+        ERR ("Requested object index does not exist.");
+        CloseSwmmOutFile();
+        return data;
+    }
+
+    if (!is_var_idx_okay)
+    {
+        ERR ("Requested variable does not exist.");
+        CloseSwmmOutFile();
+        return data;
+    }
+
+    INFO ("Fetching \"%s\"-data...", getArrayName(obj_type, var_idx, SWMM_Npolluts).c_str());
+    std::size_t const n_time_steps (static_cast<std::size_t>(SWMM_Nperiods));
+    for (std::size_t i=0; i<n_time_steps; ++i)
+    {
+        float val;
+        GetSwmmResult(obj_type_id , obj_idx, var_idx, i, &val);
+        data.push_back(static_cast<double>(val));
+    }
+
+    CloseSwmmOutFile();
+    return data;
+}
+
+std::string SwmmInterface::getArrayName(SwmmObject obj_type, std::size_t var_idx) const
+{
+    std::string const outfile (_base_name + ".out");
+    if (OpenSwmmOutFile(const_cast<char*>(outfile.c_str())) != 0)
+        return std::string("");
+
+    std::string const name = getArrayName(obj_type, var_idx, SWMM_Npolluts);
+    CloseSwmmOutFile();
+    return name;
+}
+
+std::string SwmmInterface::getArrayName(SwmmObject obj_type, std::size_t var_idx, std::size_t n_pollutants) const
+{
+    std::size_t const n_vars (0);
+    if (obj_type == SwmmObject::SUBCATCHMENT)
+    {
+        if (var_idx < n_obj_params[0])
```

Comment:
Maybe it would be better to have a variable n_obj_params_subcatchment etc per object type.